src/storage/files.c

```diff
@@ -371,32 +510,37 @@ static int load_pkcs12(BIO* in, OPTIONAL const char* desc, OPTIONAL pem_password
     /* See if an empty password will do */
     if(PKCS12_verify_mac(p12, "", 0) not_eq 0 or PKCS12_verify_mac(p12, 0, 0) not_eq 0)
     {
-        LOG(FL_WARN, "Unencrypted PKCS#12 file%s%s", for_str, desc_str);
-        pass = "";
+        LOG(FL_WARN, "Rejecting unencrypted PKCS#12 file%s%s", for_str, desc_str);
+        return 0;
     }
-    else
+    if(pem_cb is_eq 0)
     {
-        if(pem_cb is_eq 0)
-        {
-            pem_cb = password_callback;
-        }
-        len = pem_cb(tpass, PEM_BUFSIZE, 0, cb_data);
-        if(len < 0)
-        {
-            LOG(FL_ERR, "passphrase callback error for %s", desc not_eq 0 ? desc : "PKCS#12 file");
-            goto die;
-        }
-        if(len < PEM_BUFSIZE)
-        {
-            tpass[len] = 0;
-        }
-        if(0 is_eq PKCS12_verify_mac(p12, tpass, len))
-        {
-            LOG(FL_ERR, "mac verify error (wrong password?) in PKCS12 file%s%s", for_str, desc_str);
-            goto die;
-        }
-        pass = tpass;
+        pem_cb = password_callback;
+    }
+    len = pem_cb(tpass, PEM_BUFSIZE, 0, cb_data);
+    if(len < 0)
+    {
+        LOG(FL_ERR, "passphrase callback error for %s", desc not_eq 0 ? desc : "PKCS#12 file");
+        goto die;
     }
+    if(len < PEM_BUFSIZE)
+    {
+        tpass[len] = 0;
+    }
+    if(0 is_eq PKCS12_verify_mac(p12, tpass, len))
+    {
+        LOG(FL_ERR, "mac verify error (wrong password?) in PKCS12 file%s%s", for_str, desc_str);
+        goto die;
+    }
+
+    pass = tpass;
+
+    if(0 is_eq algo_allowed(p12, pass, len, NID_aes_256_cbc))
```

Comment:
Here you hard-coded `NID_aes_256_cbc`, but our PR description says
"the same algorithm that is used by "storage" functions (e.g. CREDENTIALS_save_dv)".
Instead, you should define a symbolic name for this algorithm and use it in the code where appropriate in order to prevent (future) inconsistencies.

Reply:
The algo check must not be made mandatory because otherwise general usability of PKCS#12 loading is hampered.